{"cells":[{"source":"<a href=\"https://www.kaggle.com/code/oludelehalleluyah/viral-genome-analysis?scriptVersionId=157347967\" target=\"_blank\"><img align=\"left\" alt=\"Kaggle\" title=\"Open in Kaggle\" src=\"https://kaggle.com/static/images/open-in-kaggle.svg\"></a>","metadata":{},"cell_type":"markdown"},{"cell_type":"markdown","id":"44ae689b","metadata":{"_cell_guid":"45e12bbf-af4f-4e9d-a659-9b4612021260","_uuid":"4d0d3f39-e496-4611-bf46-c6f202bcc988","papermill":{"duration":0.004798,"end_time":"2024-01-01T21:20:58.546567","exception":false,"start_time":"2024-01-01T21:20:58.541769","status":"completed"},"tags":[]},"source":["## Viral Genome Analysis Using Command Line Tools.\n","### Author: Halleluyah Darasimi Oludele\n","\n","This notebook was prepared based on the things from the Bioinformatics for Biologists: Analysing and Interpreting Genomic Datasets - Week 1 by Wellcome Connecting Science on FutureLearn. this will give full introduction to command line tools and how you can follwo through.\n","\n","Notebook will be updated regularly and do well to ask questions or clarifications concerning any part of the code.\n","\n","![SARS-CoV-2 Virus](https://cdn.who.int/media/images/default-source/mca/mca-covid-19/coronavirus-2.tmb-1024v.jpg?sfvrsn=4dba955c_18%201024w)\n","\n","Important outputs are added to the notebook in the dataset folder, so you can check through to see if yours is th edesired output."]},{"cell_type":"markdown","id":"ec4f7c72","metadata":{"papermill":{"duration":0.003841,"end_time":"2024-01-01T21:20:58.554628","exception":false,"start_time":"2024-01-01T21:20:58.550787","status":"completed"},"tags":[]},"source":["## Introduction\n","Bioinformatics tools used for genomic analysis are mostly commandline tools and require a linux terminal to work. To work with these tools, you need the following:\n","- **Operating System**: Ideally, you need a linux system to work with these tools. You can decide to use Ubuntu if you want a full linux operating system, or dual boot your windows system. For windows users, you can use Windows Subsystem for Linux (WSL). You will have full access to a linux terminal to run all these commands, but you won't have access to a GUI. Mac Users can follow along, with extra twists at some point. Another option is virtualization - using a virtual machine. \n","\n","Follow [this link](https://ubuntu.com/download/desktop) on how to install Ubuntu Linux. I strongly recommend Ubuntu 22.04 as it has long term support. \n","\n","Follow [this link](https://learn.microsoft.com/en-us/windows/wsl/install) to install WSL on your windows laptop.\n","\n","- Package manager - You will need a package manager for the installation of the packages and also for creating and managing environments. A good one is Conda. Rather than installing the full taste of conda, you can install Miniconda. This will give you the basic functionality of cona without installing all the extras."]},{"cell_type":"markdown","id":"8882ce59","metadata":{"_cell_guid":"45fb9a3f-1a84-4478-90aa-1a6b25252492","_uuid":"2f2fcfd3-b527-4569-8b48-df3294a72d6f","papermill":{"duration":0.003637,"end_time":"2024-01-01T21:20:58.562198","exception":false,"start_time":"2024-01-01T21:20:58.558561","status":"completed"},"tags":[]},"source":["## Setting up the environment using Miniconda"]},{"cell_type":"markdown","id":"a43f0e4d","metadata":{"papermill":{"duration":0.003699,"end_time":"2024-01-01T21:20:58.570056","exception":false,"start_time":"2024-01-01T21:20:58.566357","status":"completed"},"tags":[]},"source":["### Installing MiniConda\n","To install Miniconda you need to download it using wget or curel Ensure that they are installed. Then run the following commands.\n","\n","    wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh\n","    curl -O https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh\n","\n","Either should get the job done. After the download is completed, you will need to install the installation script. This will be done by running this command.\n","    \n","    bash Miniconda3-latest-Linux-x86_64.sh \n","\n","### Reading and answering the prompt when needed.\n","Answer yes when prompted and finish the installation.\n","Then choose the location where you Miniconda to be installed, then follow through and boom!!\n","\n","\n","\n","### Verifying the Installation\n","TO verify the installation, you can run the following commands\n","\n","    conda --version\n","    conda list\n","\n","\n","\n","Two things can either happen. \n","1. You can have a successful installation in which you will see the version  of conda you have installed and you will also see a list of packages installed in the enfironment by conda.\n","2. You can have a conda not found error. This means that cinda has not been added to path and to use conda, you will either have to run conda by stating the full path to where it is located or add conda to your system path so that you can access it anywhere on your system. THe secind option is what we will follow.\n","\n","To add to path, run this command:\n","\n","    export PATH=~/miniconda3/bin:$PATH\n","\n","\n","With that being done, you have installed conda and you are ready to work with the bioinformatics tools for analysing and interpreting genomics dataset."]},{"cell_type":"markdown","id":"13769ffb","metadata":{"papermill":{"duration":0.003696,"end_time":"2024-01-01T21:20:58.577641","exception":false,"start_time":"2024-01-01T21:20:58.573945","status":"completed"},"tags":[]},"source":["### Creating a virtual environment for your  Programs to run in.\n","\n","Next up is to create a virtual environment. This is to keep all the packages we will be using from interacting with the general environmental. \n","\n","In the inputs folder of this workspace is an condaenv folder. This contains an MOOC.yml file, the YAML file containing the configuration for making this environment. \n","\n","To create the environment, you will need to run this command.\n","\n","    conda env create -n MOOC --file /kaggle/input/condaenv/MOOC.yml >/dev/null\n","\n","The >/dev/null is to silence the output of the run. This will install all packages and dependencies in an environment called MOOC.\n","\n","To activate this environment,  you will need to run this command.\n","\n","    conda activate MOOC\n","\n","You may get a CondaError asking you to run conda init before conda activate.\n","\n","Run:\n","\n","    conda init\n","\n","Then close and reopen your terminal. You will be able to activate your environment by running the previous command."]},{"cell_type":"markdown","id":"f2b2d285","metadata":{"papermill":{"duration":0.003661,"end_time":"2024-01-01T21:20:58.585141","exception":false,"start_time":"2024-01-01T21:20:58.58148","status":"completed"},"tags":[]},"source":["## Analysis Pipeline\n","\n","The analysis pipeline involves the following processes:\n","\n","| Step                        | Associated File Formats                                                                      |\n","|-----------------------------|--------------------------------------------------------------------------------------------|\n","| Obtaining raw sequencing reads. | FASTQ                                                                                  |\n","| Quality Control/Trimming    | FASTQ (for trimmed reads, if applicable)                                                     |\n","| Read Alignment              | Input: FASTQ <br> Output: SAM (Sequence Alignment Map) or BAM (Binary Alignment Map)        |\n","| Genomic Coverage            | BAM (Binary Alignment Map)                                                                   |\n","| SV/CNV/Variant calling      | Input: BAM (Binary Alignment Map) <br> Output: VCF (Variant Call Format) for variants, BED or other formats for SV/CNV       |\n","| Biological Interpretation   | VCF (Variant Call Format) for genetic variants, BED or similar formats for SV/CNV annotations |\n","\n","\n","We will analyse one sample from the VOI/VOC dataset. You can download the sequence data for this sample using the following command (the sample identifier below is the accession number for the data in the European Nucleotide Archive (ENA)):\n","\n","    fastq-dump --split-files ERR5743893\n","\n","This is a sample COVID 19 dataset. The reference genome we will be using is the **Wuhan-1 reference** sequence which will be used to map the **ERR5743893**. It'll be in the input folder in this workbook."]},{"cell_type":"markdown","id":"79863bf5","metadata":{"papermill":{"duration":0.003658,"end_time":"2024-01-01T21:20:58.592679","exception":false,"start_time":"2024-01-01T21:20:58.589021","status":"completed"},"tags":[]},"source":["### Sequence Quality Control\n","\n","The most common raw data output from Next-Generation Sequencing (NGS) platforms is the FASTQ file, which is a text-based format.\n","\n","A FASTQ file defines each read by 4 lines:\n","- Line 1: Read sequence identifier (encoded descriptions of instrument, lane…)\n","- Line 2: Read sequence\n","- Line 3: « + » sign (optional: « + » followed by seq identifier)\n","- Line 4: a string of ASCII characters = Quality score associated to each Read.\n","\n","![FASTQ file format](https://ugc.futurelearn.com/uploads/assets/1f/6e/large_hero_1f6ef1d4-8a38-4a2f-8f24-d850fd594431.png)\n","\n","### FastQC for quality control.\n","\n","“FastQC” aims to provide a way to do quality control checks on fastq sequence data. Quality information is contained in the fastq file that refers to the accuracy of each base call. This helps to determine any irregularities or features that may affect your results, such as adapter contamination. MultiQC is used to aggregate results from multiple FastQC runs into one single html report. For more on FastQC, check [this link](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/).\n","\n","TO run fastQC on our sample sequence and view the output, we will do the following:\n","- First, let’s create a directory to save our fastQC outputs to:\n","\n","      mkdir -p QC_Reports \n","    \n","- Quality check using FastQC\n","         \n","      fastqc ERR5743893_1.fastq ERR5743893_2.fastq --outdir QC_Reports \n","   \n"," You should see something like this:\n","![FastQC](https://ugc.futurelearn.com/uploads/assets/d9/38/large_hero_d9388a59-94a8-48cb-94f2-5925c74a2357.png)\n","- Once fastqc has finished running, change to the QC_reports directory. You’ll see the following results files:\n","![QC_Report Folder](https://ugc.futurelearn.com/uploads/assets/8d/a0/large_hero_8da0f3ff-29ad-4e8e-aa7b-417e8d67759d.png)\n"," \n"," - You will notice that there are multiple files in the QC_Report folder. THis might become overwhelming especially when we are performing QC on many files. To combine all the fastqc output into one report, we use the multiqc command. This is done using this command.\n","\n","       multiqc . \n"," \n","**Note**: This should be done in the folder that contains the QC reports. THe multiqc command automatically finds the various reports and combine them into one.\n","\n","You can then open the report in your browser. It should look like this:\n","\n","<img src=\"https://ugc.futurelearn.com/uploads/assets/cd/45/large_hero_cd4582cd-2f0d-4d77-9a1f-cd54cda4f17c.png\" alt=\"Image\" width=\"700\" height=\"700\" />"]},{"cell_type":"markdown","id":"b36211f9","metadata":{"papermill":{"duration":0.003629,"end_time":"2024-01-01T21:20:58.600147","exception":false,"start_time":"2024-01-01T21:20:58.596518","status":"completed"},"tags":[]},"source":["#### Mapping and Aligning the Sample Sequence\n","\n","Aligning sample sequences with a reference genome is a crucial step in genetic analysis, as it reveals the most likely sources of the observed sequencing reads.\n","***\n","The reference genome serves as a comprehensive blueprint of a species' genetic composition, aiding in accurate mapping by detailing genes, regulatory elements, and other genomic attributes. \n","\n","Read mapping involves crucial steps. First, the reference genome undergoes indexing, akin to creating a book index, enabling mapping algorithms like bwa or bowtie2 to align sequencing reads efficiently. This indexing, a one-time process per mapping software, prepares for subsequent read mapping. Using specialized software, read mapping generates alignments in SAM or BAM formats, the latter being a compressed binary format. Alignments are sorted by genomic location for organization and efficient downstream processing. BAM indexing then creates an essential index for the alignment file, vital for subsequent analysis and visualization tools like the Integrated Genome Viewer (IGV).\n","\n","The mapping output typically resides in SAM format, a standardized format for storing read alignments to a reference genome. SAM files consist of fixed columns and optional key:type:value tuples. BAM, an equivalent to SAM, is optimized for speed and indexing, storing read bases, base quality, and utilizing a single conventional technique for all data types (Fig 1, Fig 2).\n","\n","<img src=\"https://ugc.futurelearn.com/uploads/assets/ce/72/large_hero_ce728418-d25b-4d85-943d-ed87e6cab341.png\" alt=\"Image\" width=\"500\" height=\"500\" />\n","\n","<img src=\"https://ugc.futurelearn.com/uploads/assets/f8/31/large_hero_f8314226-810d-4b91-bdbf-05be43f1ca37.png\" alt=\"Image\" width=\"500\" height=\"500\" />\n","\n","#### Aligning the sample sequence\n","\n","For this,we will use the BWA-MEM tool.\n","- First, we will create a folder to store our output.\n","           \n","      mkdir Mapping\n","      \n","- The first step is indexing the reference genome. This makes information retrieval quicker and more effective and, lessens computational complexities and supports advanced querying methods. This is done like this:\n","\n","      bwa index MN908947.fasta\n","      \n","- Now we can map our target sample to the reference genome\n","\n","      bwa mem MN908947.fasta ERR5743893_1.fastq ERR5743893_2.fastq > Mapping/ERR5743893.sam\n","\n","- Once the bwa mem is done running we can change to the Mapping folder to see the output\n","      \n","      cd Mapping\n","      ls -lhrt\n","      \n","    You should see something like\n","\n","![this](https://ugc.futurelearn.com/uploads/assets/87/ab/large_hero_87ab98ed-b441-48ab-ad70-ecd472b162f4.png)    \n","\n","   The output of the bwa mem is a sam file which is quite large. To compress it, we can convert it to a sam file. We will use samtools for this purpose.\n","\n","        cd ..\n","        samtools view -@ 20 -S -b Mapping/ERR5743893.sam > Mapping/ERR5743893.bam\n","\n","   To get more information on the samtools being used, you can run\n","\n","        samtools -help\n","\n","   The new bam file has a size of 11 Mb compared to the sam file of 147 Mb.\n","\n","- The SAM and BAM files generated initially are sorted based on the order of the reads in the fastq files. However, it is more practical to have these files sorted based on the alignment positions in the reference genome rather than the original read order. \n","\n","- To accomplish this, we will utilize `samtools` for sorting, which enables us to arrange the reads and their corresponding mapping information in a BAM file according to their alignment positions in the reference genome.\n","\n","      samtools sort -@ 4 -o Mapping/ERR5743893.sorted.bam Mapping/ERR5743893.bam\n","  \n","  Next we index the sorted BAM file. This allows for **effective variant calling**, **visualisation**, and **genomics research**.\n","  \n","      samtools index Mapping/ERR5743893.sorted.bam\n","      \n","  This can then be visualized in the integrative genomics viewer using the bam and the index files."]},{"cell_type":"markdown","id":"fc20d617","metadata":{"papermill":{"duration":0.00365,"end_time":"2024-01-01T21:20:58.607779","exception":false,"start_time":"2024-01-01T21:20:58.604129","status":"completed"},"tags":[]},"source":["#### Variant Calling\n","\n","A variant is any change observed when comparing two sequences. The process of analysing and characterising this variants is called **variant calling**. These variants include:\n","\n","| Term            | Meaning                                          |\n","|-----------------|--------------------------------------------------|\n","| Insertions (Ins)| Addition of nucleotides compared to the reference |\n","| Deletions       | Removal of nucleotides compared to the reference |\n","| Indel           | General term for Insertions/Deletions             |\n","| SNPs            | Single Nucleotide Polymorphisms                   |\n","| CNVs            | Copy Number Variations                            |\n","| Translocations  | Movement of genetic material between chromosomes  |\n","| Inversions      | Reversal of the orientation of a DNA segment      |\n","\n","##### Factors to consider in variant calling\n","- In variant calling, ensuring accuracy involves filtering out potential false calls.\n","- False calls can result from:\n","  - Contamination during sample preparation.\n","  - Errors during PCR amplification.\n","  - Sequencing errors.\n","  - Challenges in handling homopolymer runs.\n","  - Issues with mapping to repetitive sequences.\n","  - Alignment errors.\n","- False SNPs might appear near indels, and ambiguous indel alignment can contribute to false variants.\n","- Addressing these factors is crucial for refining variant calling and obtaining reliable genetic information.\n","\n","##### Process of the variant calling\n","- First, we need to index the reference genome using samtools\n","      \n","      samtools faidx MN908947.fasta\n","\n","- Now we will use FreeBayes to identify variants in our target sequence.\n","\n","      freebayes -f MN908947.fasta Mapping/ERR5743893.sorted.bam  > ERR5743893.vcf\n","\n","- Finally it is good practice to compress the output so that it takes up less space. Also, popular tools can also work with the compressed format of the vcf file so there is no need to uncompress it later.\n","\n","      bgzip ERR5743893.vcf\n","      tabix ERR5743893.vcf.gz"]},{"cell_type":"markdown","id":"4d3bb18e","metadata":{"papermill":{"duration":0.003731,"end_time":"2024-01-01T21:20:58.615478","exception":false,"start_time":"2024-01-01T21:20:58.611747","status":"completed"},"tags":[]},"source":["## Interpreting the variant call file\n","\n","To do this, you will need to use the BCFtools. First, you need to install it.\n","\n","- For Linux users, use this command\n","\n","      sudo apt install bcftools\n","      \n","- To install in an active conda environmen, use the following command:\n","\n","      conda install -c bioconda bcftools\n","      \n","\n","- To install on Mac, use homebrew by using the following command:\n","\n","      brew install bcftools\n","      \n","### Interpreting the output of the vcf file\n","\n","To identify the variants, you can use the following commands:\n","\n","       bcftools view -o output_snps.vcf.gz -O z -v snps,indels ERR5743893.vcf.gz\n","       \n","       bcftools query -f '%TYPE\\n' ERR5743893.vcf.gz | sort | uniq -c\n","\n","The output of this is that there are **1 INDELS** and **75 SNPs**\n"," \n","For more information on how to use the bcf tools, you can access it from [here](https://samtools.github.io/bcftools/bcftools.html)"]},{"cell_type":"markdown","id":"15a40385","metadata":{"papermill":{"duration":0.003674,"end_time":"2024-01-01T21:20:58.623114","exception":false,"start_time":"2024-01-01T21:20:58.61944","status":"completed"},"tags":[]},"source":["## References\n","\n","THe images are got from the Bioinformatics for Biologists: Analysing and Interpreting genomics Dataset course on FutureLearn by the Wellcome Connecting Science."]}],"metadata":{"kaggle":{"accelerator":"none","dataSources":[{"datasetId":4239511,"sourceId":7306829,"sourceType":"datasetVersion"},{"datasetId":4247781,"sourceId":7319785,"sourceType":"datasetVersion"}],"dockerImageVersionId":30626,"isGpuEnabled":false,"isInternetEnabled":true,"language":"python","sourceType":"notebook"},"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.12"},"papermill":{"default_parameters":{},"duration":6.54874,"end_time":"2024-01-01T21:21:01.523797","environment_variables":{},"exception":null,"input_path":"__notebook__.ipynb","output_path":"__notebook__.ipynb","parameters":{},"start_time":"2024-01-01T21:20:54.975057","version":"2.4.0"}},"nbformat":4,"nbformat_minor":5}